File: Evaluations/Score_Chords.m
clear all
clc

% Directories for prediction and GT
GT_dir = '../chordlabs_minmaj/';
Predict_dir = '../Predictions/Matt_pretrained_minmaj/';

% Path for toolbox
addpath(genpath('../'))

% Is something appended to the predictions? (ie '_prediction')?
appended = '_prediction.txt';

% Get filenames
GT_files = getAllFiles(GT_dir);
Predict_files = getAllFiles(Predict_dir);

% Alphabet
alphabet = 'minmaj';

% Store results
Overlap = [];
song_lengths = [];

% Main loop
for i=1:length(GT_files)

  disp(['Evaluating song: ' num2str(i) ' of ' num2str(length(GT_files)) ': ' GT_files{i}])

  % Get expected Prediction name
  [trash,localname,extension] = fileparts(GT_files{i});
  Predict_name = [Predict_dir localname appended];

  % Read GT and Prediction
  f_GT = fopen(GT_files{i}); GT = textscan(f_GT,'%s%s%s'); fclose(f_GT);
  f_P = fopen(Predict_name); P = textscan(f_P,'%s%s%s'); fclose(f_P);

  GT_chords = GT{3};
  P_chords = P{3};

  GT_times = [];
  for t=1:length(GT{1})
    GT_times = [GT_times ; [str2num(GT{1}{t}) str2num(GT{2}{t})]]; 
  end

  P_times = [];
  for t=1:length(P{1})
    P_times = [P_times ; [str2num(P{1}{t}) str2num(P{2}{t})]];
  end

  % Fix prediction to length of GT
  if P_times(end,end) < GT_times(end,end)
    P_times = [P_times ; [P_times(end-1,end) GT_times(end,end)]];
    P_chords = [P_chords ; 'N'];
  elseif P_times(end,end) > GT_times(end,end)
    bad_rows = find(P_times(:,1) > GT_times(end,end));
    P_times(bad_rows,:) = [];
    P_chords(bad_rows) = [];
    P_times(end,end) = GT_times(end,end); 
  else
    % Fine
  end

  % Sample so they're the same alphabet
  switch alphabet 
    case 'minmaj'
      GT_chords = reduce_to_minmaj(GT_chords);
      P_chords = reduce_to_minmaj(P_chords);
    case 'triads'
      GT_chords = reduce_to_triads(GT_chords);
      P_chords = reduce_to_triads(P_chords);
    case 'quads'
      GT_chords = reduce_to_quads(GT_chords);
      P_chords = reduce_to_quads(P_chords);
  end
 
  % Now sample each chord at 1 kHz
  unique_chords = unique([GT_chords P_chords]);
  GT_chord_sample = [];
  P_chord_sample = [];
  for t=1:length(GT_chords)
    sym = strmatch(GT_chords{t},unique_chords);
    duration = round(1000*(GT_times(t,2) - GT_times(t,1)));
    GT_chord_sample = [GT_chord_sample repmat(sym,[1, duration])];
  end

  for t=1:length(P_chords)
    sym = strmatch(P_chords{t},unique_chords);
    duration = round(1000*(P_times(t,2) - P_times(t,1)));
    P_chord_sample = [P_chord_sample repmat(sym,[1, duration])];
  end

  % Still can have rounding effects, but only a maximum of 
  % nchords/1000 seconds...
  minlen = min([length(P_chord_sample) length(GT_chord_sample)]);
  GT_chord_sample = GT_chord_sample(1:minlen);
  P_chord_sample = P_chord_sample(1:minlen);

  % Finally, output score
  Overlap(i) = 100*mean(eq(GT_chord_sample,P_chord_sample));
  song_lengths(i) = minlen;
end

song_lengths = song_lengths/sum(song_lengths);

disp('***********')
disp(['Mean Overlap: ' num2str(mean(Overlap)) '%'])
disp(['Total Overlap: ' num2str(song_lengths*Overlap') '%'])
disp('***********')

File: Evaluations/Python_Eval/Score_chords.py
# Directories for prediction and GT
GT_dir = '/Users/mattmcvicar/Desktop/Work/New_chroma_features/Package/chordlabs/'
Predict_dir = '/Users/mattmcvicar/Desktop/Work/New_chroma_features/Package/Predictions/Matt_pretrained_complex/'

# Is something appended to the predictions? (ie '_prediction')?
appended = '_prediction.txt'

# Get filenames
import os
GT_files = os.listdir(GT_dir)
Predict_files = os.listdir(Predict_dir)

# Only process the right file types
file_ext = '.lab'
GT_files = [f for f in GT_files if os.path.splitext(f)[1] == file_ext]
Predict_files = [f for f in Predict_files if os.path.splitext(f)[1] == file_ext]

# Alphabet
alphabet = 'minmaj';

# Store results
CP_Overlap = []
NP_Overlap = []
MIREX_Overlap = []
song_lengths = []

# Ignore these files
ignore = [150] # Revolution 9

# Main loop
import numpy as np
import reduce_chords
for (index,GT_file) in enumerate(GT_files):

  print 'Evaluating song: ' + str(index + 1) + ' of ' + str(len(GT_files)) + ': ' + GT_file

  # Get expected Prediction name
  localname,extension = os.path.splitext(GT_file)
  Predict_name = Predict_dir + localname + appended

  # Read GT and Prediction
  GT = open(GT_dir + GT_file).readlines()
  P = open(Predict_name).readlines()

  GT_chords = []
  GT_start_times = []; GT_end_times = []
  for line in GT:
    start_time,end_time,chord = line.split() 
    
    GT_chords.append(chord)
    GT_start_times.append(np.double(start_time)) 
    GT_end_times.append(np.double(end_time)) 

  P_chords = []
  P_start_times = []; P_end_times = []
  for line in P:
    start_time,end_time,chord = line.split() 
    
    P_chords.append(chord)
    P_start_times.append(np.double(start_time)) 
    P_end_times.append(np.double(end_time)) 

  # Fix prediction to length of GT
  if P_end_times[-1] < GT_end_times[-1]:
    P_start_times.append(P_start_times[-1])  
    P_end_times.append(GT_end_times[-1])
    P_chords.append('N')
  elif P_end_times[-1] > GT_end_times[-1]:  
    remove_rows = [i for (i,t) in enumerate(P_start_times) if t > GT_end_times[-1] or P_end_times[i] > GT_end_times[-1]]
    
    # Remove them
    del P_start_times[remove_rows[0]:]
    del P_end_times[remove_rows[0]:]
    del P_chords[remove_rows[0]:]
  else:
    # Fine
    pass

  # Sample so they're the same alphabet 
  GT_chords_reduce = reduce_chords.reduce_chords(GT_chords,alphabet)
  P_chords_reduce = reduce_chords.reduce_chords(P_chords,alphabet)
  
  # Get the bass notes from the full chord labels
  GT_bass = reduce_chords.reduce_chords(GT_chords,'bass')
  P_bass = reduce_chords.reduce_chords(P_chords,'bass')

  # Now sample each chord at 1 kHz. Store pitch classes and sorted pitch classes
  # for NP and CP
  unique_chords = list(set(np.hstack((GT_chords_reduce,P_chords_reduce))))
  GT_chord_sample_CP = []; GT_chord_sample_NP = []; GT_chord_sample_MIREX = []
  P_chord_sample_CP = []; P_chord_sample_NP = []; P_chord_sample_MIREX = []
  
  for c_index,chord in enumerate(GT_chords_reduce):
    duration = int(round(1000*(GT_end_times[c_index] - GT_start_times[c_index])))
    
    # Need bass symbol for CP, first two notes for MIREX
    sym = reduce_chords.chord2pitchclasses(chord)[0]
    bass_sym = GT_bass[c_index]
    MIREX_sym = reduce_chords.chord2pitchclasses(GT_chords[c_index])[0]
    if len(MIREX_sym) < 2:
      pass
    else:
      MIREX_sym = MIREX_sym[:2]
        
    GT_chord_sample_NP.extend([sym]*duration)
    GT_chord_sample_CP.extend([[sym, bass_sym]]*duration)
    GT_chord_sample_MIREX.extend([MIREX_sym]*duration)
    
  for c_index,chord in enumerate(P_chords_reduce):
    duration = int(round(1000*(P_end_times[c_index] - P_start_times[c_index])))
    
    # Need bass symbol for CP, first two notes for MIREX
    sym = reduce_chords.chord2pitchclasses(chord)[0]
    bass_sym = P_bass[c_index]
    MIREX_sym = reduce_chords.chord2pitchclasses(P_chords[c_index])[0]
    if len(MIREX_sym) < 2:
      pass
    else:
      MIREX_sym = MIREX_sym[:2]
        
    P_chord_sample_NP.extend([sym]*duration)
    P_chord_sample_CP.extend([[sym, bass_sym]]*duration)
    P_chord_sample_MIREX.extend([MIREX_sym]*duration)
    
  # Still can have rounding effects, but only a maximum of 
  # nchords/1000 seconds...
  minlen = np.min([len(P_chord_sample_CP),len(GT_chord_sample_CP)])
  GT_chord_sample_CP = GT_chord_sample_CP[:minlen]; 
  P_chord_sample_CP = P_chord_sample_CP[:minlen]
  
  GT_chord_sample_NP = GT_chord_sample_NP[:minlen]
  P_chord_sample_NP = P_chord_sample_NP[:minlen]
 
  GT_chord_sample_MIREX = GT_chord_sample_MIREX[:minlen]
  P_chord_sample_MIREX = P_chord_sample_MIREX[:minlen]
  
  # Display?
  # import matplotlib.pyplot as plt
  # plt.imshow(np.vstack((np.array(GT_chord_sample),np.array(P_chord_sample))),interpolation="nearest",aspect="auto"); plt.show() 
  
  # CP = chords have the same notes
  
  # Finally, output score
  CP_correct = [GT_chord_sample_CP[i] == p for (i,p) in enumerate(P_chord_sample_CP)]
  NP_correct = [GT_chord_sample_NP[i] == p for (i,p) in enumerate(P_chord_sample_NP)]
  MIREX_correct = [GT_chord_sample_MIREX[i] == p for (i,p) in enumerate(P_chord_sample_MIREX)]
  
  CP_Overlap.append(100*np.mean(CP_correct))
  NP_Overlap.append(100*np.mean(NP_correct))
  MIREX_Overlap.append(100*np.mean(MIREX_correct))
  song_lengths.append(minlen)
  
# Remove some songs
CP_Overlap = [o for index,o in enumerate(CP_Overlap) if index not in ignore]
NP_Overlap = [o for index,o in enumerate(NP_Overlap) if index not in ignore]
MIREX_Overlap = [o for index,o in enumerate(MIREX_Overlap) if index not in ignore]

song_lengths = [s for index,s in enumerate(song_lengths) if index not in ignore]  
  
# Normalise song lengths  
song_lengths = np.true_divide(song_lengths,np.sum(song_lengths))

print '***************************'
print 'Mean Chord overlap: ' + str(round(np.mean(CP_Overlap),2)) + '%'
print 'Mean Note overlap: ' + str(round(np.mean(NP_Overlap),2)) + '%'
print 'Mean MIREX overlap: ' + str(round(np.mean(MIREX_Overlap),2)) + '%'
print 'Total Chord overlap: ' + str(round(np.dot(song_lengths,CP_Overlap),2)) + '%'
print 'Total Note overlap: ' + str(round(np.dot(song_lengths,NP_Overlap),2)) + '%'
print 'Total MIREX overlap: ' + str(round(np.dot(song_lengths,MIREX_Overlap),2)) + '%'
print '***************************'

